### PR TITLE
Support RPC server and client functions

### PR DESCRIPTION
**Pull request by LiZhang-2020**
jsonrpc: support RPC server and client functions

Support basic Json RPC functions,
such as RPC server and client.
Depends on libev http://software.schmorp.de/pkg/libev.html

examples/vdpa: support RPC message and handle

Create VDPA RPC thread to handle RPC message from App.

### CODE REVIEW

File: lib/jsonrpc/cJSON.c [hunk 1]
```diff
@@ -0,0 +1,1049 @@
...
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
```
Comment (kailiangz1) [inline lib/jsonrpc/cJSON.c]: like this:
/* SPDX-License-Identifier: BSD-3-Clause
 * Copyright (c) 2022 NVIDIA Corporation & Affiliates
 */

File: examples/vdpa/vdpa_rpc.c [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+	cJSON *pf_dev = cJSON_GetObjectItem(params, "dev");
+	cJSON *result = NULL;
+	struct vdpa_rpc_context *rpc_ctx;
+
+	RTE_LOG(ERR, RPC, "lizh cJSON mgmtpf...start \n");
+	rpc_ctx = (struct virtnet_rpc_context *)ctx->data;
+	pthread_mutex_lock(&rpc_ctx->rpc_lock);
+	if (pf_add && pf_dev) {
+		/* Parse PCI device name*/
+		RTE_LOG(ERR, RPC, "lizh cJSON mgmtpf.pf_add %s:%d pf_dev %s: %s...start \n", pf_add->string, pf_add->valueint, pf_dev->string, pf_dev->valuestring);
```
Comment (kailiangz1) [inline examples/vdpa/vdpa_rpc.c]: lizh should not here

Comment (kailiangz1) [inline examples/vdpa/vdpa_rpc.c]: one line should not exceed 80 character

File: examples/vdpa/main.c [hunk 3]
```diff
@@ -529,6 +533,7 @@ main(int argc, char *argv[])
 	struct rte_device *dev;
 	struct rte_dev_iterator dev_iter;
 
+	vdpa_rpc_start(&vdpa_rpc_ctx);
```
Comment (kailiangz1) [inline examples/vdpa/main.c]: judge return value

File: examples/vdpa/vdpa_rpc.c [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+static cJSON *mgmtpf(jrpc_context *ctx, cJSON *params, cJSON *id)
+{
+	cJSON *pf_add = cJSON_GetObjectItem(params, "add");
+	cJSON *pf_remove = cJSON_GetObjectItem(params, "remove");
+	cJSON *pf_list = cJSON_GetObjectItem(params, "list");
+	cJSON *pf_dev = cJSON_GetObjectItem(params, "dev");
+	cJSON *result = NULL;
+	struct vdpa_rpc_context *rpc_ctx;
+
+	RTE_LOG(ERR, RPC, "lizh cJSON mgmtpf...start \n");
```
Comment (kailiangz1) [inline examples/vdpa/vdpa_rpc.c]: INFO

File: examples/vdpa/vdpa_rpc.c [hunk 5]
```diff
@@ -0,0 +1,251 @@
...
+	rte_cpuset_t cpuset;
+	pthread_attr_t attr;
+	char name[32];
+	int ret;
+
+	RTE_LOG(ERR, RPC, "lizh mlx5_vdpa_rpc_start...start \n");
+	pthread_attr_init(&attr);
+	ret = pthread_attr_setschedpolicy(&attr, SCHED_RR);
+	if (ret) {
+		RTE_LOG(ERR, RPC, "Failed to set thread sched policy = RR.\n");
```
Comment (kailiangz1) [inline examples/vdpa/vdpa_rpc.c]: print ret value and return ret

File: examples/vdpa/vdpa_rpc.h [hunk 6]
```diff
@@ -0,0 +1,40 @@
+/* SPDX-License-Identifier: BSD-3-Clause
+ * Copyright 2022, NVIDIA CORPORATION & AFFILIATES.
+ */
+
+#ifndef _VHOST_BLK_H_
```
Comment (kailiangz1) [inline examples/vdpa/vdpa_rpc.h]: _VDPA_RPC_H_

File: examples/vdpa/vdpa_rpc.h [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+#define _VHOST_BLK_H_
+
+#include <getopt.h>
+#include <signal.h>
+#include <stdint.h>
+#include <string.h>
+#include <unistd.h>
+#include <sys/eventfd.h>
+
+#include <rte_vhost.h>
```
Comment (kailiangz1) [inline examples/vdpa/vdpa_rpc.h]: why need this 2 header file?
vhost.h eventfd.h

**Review by kailiangz1 — CHANGES_REQUESTED** [review]

File: lib/jsonrpc/version.map [hunk 8]
```diff
@@ -9,6 +9,7 @@ DPDK_22 {
 	jrpc_register_procedure;
 	jrpc_deregister_procedure;
 	jsonrpc_client_call_method;
+	rte_rpc_vdpa_global_ops_init;
```
Comment (kailiangz1) [inline lib/jsonrpc/version.map]: who will call rte_rpc_vdpa_global_ops_init?

File: lib/jsonrpc/cJSON.c (hunk 1 above)
Comment (LiZhang-2020) [reply lib/jsonrpc/cJSON.c]: This file is coded by Henirgue from open source and copyright is not belong to NVIDIA.
Such as:
Copyright (c) 2012-2013 Henrique Gomes

File: examples/vdpa/vdpa_rpc.c (hunk 2 above)
Comment (LiZhang-2020) [reply examples/vdpa/vdpa_rpc.c]: Will deleted all log with lizh.

Comment (LiZhang-2020) [reply examples/vdpa/vdpa_rpc.c]: Done

File: lib/jsonrpc/version.map (hunk 8 above)
Comment (LiZhang-2020) [reply lib/jsonrpc/version.map]: Admin queue will use it to register the PF driver ops functions.

File: examples/vdpa/main.c (hunk 3 above)
Comment (LiZhang-2020) [reply examples/vdpa/main.c]: Done

File: examples/vdpa/vdpa_rpc.c (hunk 4 above)
Comment (LiZhang-2020) [reply examples/vdpa/vdpa_rpc.c]: Done

File: examples/vdpa/vdpa_rpc.c (hunk 5 above)
Comment (LiZhang-2020) [reply examples/vdpa/vdpa_rpc.c]: Done

File: examples/vdpa/vdpa_rpc.h (hunk 6 above)
Comment (LiZhang-2020) [reply examples/vdpa/vdpa_rpc.h]: Done

File: examples/vdpa/vdpa_rpc.h (hunk 7 above)
Comment (LiZhang-2020) [reply examples/vdpa/vdpa_rpc.h]: Delete them

File: examples/vdpa/main.c [hunk 9]
```diff
@@ -549,20 +577,6 @@ main(int argc, char *argv[])
 		cmdline_interact(cl);
 		cmdline_stdin_exit(cl);
 	} else {
-		RTE_DEV_FOREACH(dev, "class=vdpa", &dev_iter) {
```
Comment (kailiangz1) [inline examples/vdpa/main.c]: should not remove prvious logic.

File: examples/vdpa/vdpa_rpc.c [hunk 10]
```diff
@@ -0,0 +1,358 @@
...
+	if (pf_add && pf_dev) {
+		/* Parse PCI device name*/
+		result = vdpa_pf_dev_add(pf_dev->valuestring);
+	}
+	if (pf_remove && pf_dev) {
+		/* Parse PCI device name*/
+		result = vdpa_pf_dev_remove(pf_dev->valuestring);
+	}
+	if (pf_list) {
+		result = vdpa_pf_dev_list();
```
Comment (kailiangz1) [inline examples/vdpa/vdpa_rpc.c]: else if{
else if
they can't true in the same time ?

**Review by kailiangz1 — CHANGES_REQUESTED** [review]

File: examples/vdpa/main.c (hunk 9 above)
Comment (LiZhang-2020) [reply examples/vdpa/main.c]: Done

File: examples/vdpa/vdpa_rpc.c (hunk 10 above)
Comment (LiZhang-2020) [reply examples/vdpa/vdpa_rpc.c]: Done

File: lib/jsonrpc/rte_rpc.h [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+	uint64_t features;
+	uint32_t mtu;
+	struct rte_ether_addr mac;
+};
+
+enum vdpa_vf_modify_flags {
+	VDPA_VF_MSIX_NUM,
+	VDPA_VF_QUEUE_NUM,
+	VDPA_VF_QUEUE_SZIE,
+	VDPA_VF_NET_FEATURES,
```
Comment (GavinLi-NV) [inline lib/jsonrpc/rte_rpc.h]: Remove "_NET"

Comment (LiZhang-2020) [reply lib/jsonrpc/rte_rpc.h]: Done

File: examples/vdpa/main.c [hunk 12]
```diff
@@ -92,10 +96,10 @@ parse_args(int argc, char **argv)
 		}
 	}
 
-	if (iface[0] == '\0' && interactive == 0) {
+	/*if (iface[0] == '\0' && interactive == 0) {
```
Comment (GavinLi-NV) [inline examples/vdpa/main.c]: Still temporary code?

File: examples/vdpa/vdpa_rpc.h [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+
+#define MAX_PATH_LEN 128
+/* VDPA RPC */
+#define VIRTNET_FEATURE_SZ	64
+#define MAX_VDPA_SAMPLE_PORTS 1024
+#define RTE_LOGTYPE_RPC RTE_LOGTYPE_USER1
+#define VDPA_LOCAL_HOST "127.0.0.1"
+
+enum {
+	VDPA_RPC_PORT = 12190,
```
Comment (GavinLi-NV) [inline examples/vdpa/vdpa_rpc.h]: Could user change this in runtime?

File: lib/jsonrpc/version.map [hunk 14]
```diff
@@ -0,0 +1,15 @@
+DPDK_22 {
```
Comment (GavinLi-NV) [inline lib/jsonrpc/version.map]: maybe we should use "INTERNAL" instead?

File: examples/vdpa/main.c (hunk 12 above)
Comment (LiZhang-2020) [reply examples/vdpa/main.c]: Delete it

File: examples/vdpa/vdpa_rpc.h (hunk 13 above)
Comment (LiZhang-2020) [reply examples/vdpa/vdpa_rpc.h]: No, it is hardcode. It is exampler for customer. They can change it in his code

File: lib/jsonrpc/version.map (hunk 14 above)
Comment (LiZhang-2020) [reply lib/jsonrpc/version.map]: Done

File: lib/jsonrpc/rte_rpc.h [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+	uint32_t queue_size;
+	uint64_t features;
+	uint32_t mtu;
+	struct rte_ether_addr mac;
+};
+
+enum vdpa_vf_modify_flags {
+	VDPA_VF_MSIX_NUM,
+	VDPA_VF_QUEUE_NUM,
+	VDPA_VF_QUEUE_SZIE,
```
Comment (GavinLi-NV) [inline lib/jsonrpc/rte_rpc.h]: Typo

Comment (LiZhang-2020) [reply lib/jsonrpc/rte_rpc.h]: Done

**Pull request merged**